event-bus/api/src/main/java/org/apache/james/events/EventBus.java

```diff
@@ -54,6 +55,13 @@ default Publisher<Registration> register(EventListener listener, RegistrationKey
 
     Publisher<Registration> register(EventListener.ReactiveEventListener listener, RegistrationKey key);
 
+    default Publisher<Registration> register(EventListener.ReactiveEventListener listener, Collection<RegistrationKey> keys) {
+        return Flux.fromIterable(keys)
+            .concatMap(key -> register(listener, key))
+            .reduce((reg1, reg2) -> () -> Mono.from(reg1.unregister())
+                .then(Mono.from(reg2.unregister())));
+    }
```

Comment:
IMO we could / should use Flux.merge.

There's no need to delay `reg2.unregister()` subscription.

```
return Flux.fromIterable(keys)
            .concatMap(key -> register(listener, key))
            .collectList()
            .flatMap(regs -> Flux.merge(regs).then());
```

?

Reply:
Your proposal does not work exactly but I take your idea and mutate it a bit. Please check my latest fixup commit.